### DISCUSSION TITLE
资源文件在磁盘中布局的问题

### DISCUSSION THREAD

**Discussion by rexzhang** [py-pal/py-pal#3] [Ideas]
# 资源文件布局一般来说不外乎两种模式

- 使用多层文件夹的方式分类存放(res/music,res/video,res/ui);后面简称「文件夹」模式
- 使用一种所谓的「资源包」模式;就是使用一种特别定义的文件格式,将众多文件以虚拟的目录树的方式存放于一个单元文件当中;有点类似 `.zip` 等压缩文件包

# 优劣分析:

## 文件夹
- 无直接访问方面的性能问题
- 开发阶段友好
- 如果文件系统缓存逻辑不佳,或者目标计算机内存不足,会有同一个目录下文件过大导致的查找性能问题
- 对分发不友好
  - 大量资源文件迭代的情况下,容易出现分发错误问题(人的因素)
  - 更新分发时无法使用二进制 diff 来快速的完成更新,更新程序需要了解资源文件结构
- 在大容量磁盘(存储空间)环境下,存在严重的实际磁盘容量浪费问题

## 资源包
- 因为存在重新定位(seek)的问题,有访问上的性能损失
- 需要整个程序在所有资源的代码均支持以流(steam)/句(handle)柄方式打开资源文件
- 需要解决开发阶段快速访问临时添加的资源文件的问题,这个需要在适配层做更多的实现

### 自定义
可以参考暴雪的 `.mpq`, 设计上非常完善了

### ZIP
- 这是最简单直粗暴的方式
- 可以使用 `zipfile` 这个库
- 可以根据需要对单独的某个资源包开启不同的压缩级别(zipfile.ZIP_STORED为不压缩)
- 需要使用 `ZipFile.open` 来打开文件
- 为了解决同时访问本地文件夹资源,需要实现一个适配层来接管所有文件类操作

**Comment by rexzhang** [comment]
我倾向的方案是
- zip/zip-patch/本地文件 三层叠加,后面的权重大
- zip: 允许同时存在多个,适配层自动拼接
- zip-patch: 就是在自动拼接的前提下,允许有基于优先级的覆盖逻辑
- 本地文件: 本地可以有一个唯一的文件夹,里面内容可以叠加/覆盖进去

**Comment by alanshi** [comment]
> zip/zip-patch/本地文件 三层叠加,后面的权重大

这个方案比较可行，不过有些细节问题：

1. 分析仙剑原有的资源包，并对解析出来的资源进行重新打包，需要按什么样的粒度去划分（例如某个资源包内包含有不同场景的几张图片，是沿用这个资源包的原有结构进行打包还是用新的命名规则打包）
2. 某个zip包的单个资源命名是xxx.jpg/xxx.bmp 还是别的格式?是否意味着zip包解压后，资源就可以作为文件直接使用

**Comment by rexzhang** [comment]
首先,我们确定一个完整的虚拟的资源文件树.这个树的根在本地文件夹中是 `res/` 目录,在多个 zip 中的资源虚拟解开后均平铺的放在 `res/` 下.实际调用的时候不用带 `res` 这个前缀

针对原有资源包解析出来的资源,如何处理,我觉得要先看原版资源包内是否有虚拟的目录结构,我盲猜应该有.我们可以将其内部资源放到目录 `res/pal1w95`,在使用上也尊重原有的目录结构(完全没有子目录也是一组目录结构)

我们对原版资源文件的替换,估计会从 UI 部分开始;而我们自己重新生成/引入的资源可以按照资源类别分类存储管理,这个每一种不同的资源都可以有自己的目录规则

最后整个虚拟的文件树结构大概是
```
res
+pal1w95
+ui
+texture
...
+sounds
+music
+video
``` 

针对在我们自己的资源库中图片等资源文件应该用什么格式的问题:
- 原版的资源包内存储的格式应该不是 BMP/PNG 之类的格式,应该是某种自定义的类 BMP 格式,因为标准的图片/音视频文件都带有文件头信息,在游戏中是没有存在意义的,因为,游戏的某一种资源肯定是约定俗成的某个特别格式.移除掉千篇一律的文件头可以省下大量的存储空间
- 在资源库中使用什么格式保存应该遵循代码中如何使用最省事,执行效率最高来确定.
- 我的想法是:先确定各自资源的使用方式,再定存储格式
- 至于默认的导出格式:
  - 原版资源包中的图片等资源应该是使用了一些特别的规则来解决 BMP 中大量留白区域占据存储空间的问题,这样对调用非常不友好
  - 考虑到现代计算机硬件的性能(CPU/RAM)的提升,很多时候可以考虑使用压缩的格式,特别是对不常驻内存的资源
  - 图片
    - 高频调用/常驻内存 bmp
    - 低频调用 png
  - 音频
    - 音效 wav
    - 背景音乐,语音 mp3/ogg
    - [pyogg 的 macOS 包还不支持 wheel 发布](https://github.com/TeamPyOgg/PyOgg/issues/32),所以作为备选格式
  - 视频类
    - mp4

**Comment by rexzhang** [comment]
考虑到 arcade 使用图片资源的方式,磁盘中都是用 png 格式也没啥问题,顶多读取是可能会慢一点,完全可以考虑先都转换为 png,以后看情况再调整
